first version of schema design essentials (#144)
questdb/documentation: documentation/guides/schema-design-essentials.md

---
title: Schema Design Essentials

> slug: /schema-design-essentials

Reviewer: ```suggestion
slug: schema-design-essentials
```

description:
  Learn how to design efficient schemas in QuestDB. This guide covers best practices for partitioning, indexing, symbols, timestamps, deduplication, retention strategies, and schema modifications to optimize performance in time-series workloads.
---

> # Schema Design in QuestDB

Reviewer: ```suggestion
```

This guide covers key concepts and best practices to take full advantage of QuestDB’s performance-oriented architecture, highlighting some important differences with most databases.

## QuestDB’s Single Database Model

> QuestDB has a **single database per instance**. Unlike PostgreSQL and other database engines, where you may have multiple databases or multiple schemas within an instance, in QuestDB, you operate within a single namespace. The default database is named `qdb`, and this can be changed via configuration. However, there is no need to issue `USE DATABASE` commands—once connected, you can immediately start querying and inserting data.

Reviewer: ```suggestion
QuestDB has a **single database per instance**. Unlike PostgreSQL and other database engines, where you may have multiple databases or multiple schemas within an instance, in QuestDB, you operate within a single namespace. 

The default database is named `qdb`, and this can be changed via configuration. However, unlike a standard SQL database, there is no need to issue `USE DATABASE` commands—once connected. You can immediately start querying and inserting data.
```

### Multi-Tenancy Considerations

> If you need **multi-tenancy**, you will need to manage table names manually, often by using **prefixes** for different datasets. Since QuestDB does not support multiple schemas, this is the primary way to segment data. In QuestDB Enterprise, you can enforce permissions per table to restrict access, allowing finer control over multi-tenant environments.

Reviewer: ```suggestion
If you need **multi-tenancy**, you will need to manage table names manually, often by using **prefixes** for different datasets. Since QuestDB does not support multiple schemas, this is the primary way to segment data. In QuestDB Enterprise, you can [enforce permissions per table to restrict access](/docs/operations/rbac/), allowing finer control over multi-tenant environments.
```

## PostgreSQL Protocol Compatibility

QuestDB is **not** a PostgreSQL database but is **compatible with the PostgreSQL wire protocol**. This means you can connect using PostgreSQL-compatible libraries and clients and execute SQL commands. However, compatibility with PostgreSQL system catalogs, metadata queries, data types,
and functions is limited.

While most PostgreSQL compatible low-level libraries will work with
QuestDB, some higher level components depending heavily on PostgreSQL metadata might fail. If you find one of such use cases, please do report it as an [issue on GitHub](https://github.com/questdb/questdb/issues) so we can track it.

## Creating a Schema in QuestDB

### Recommended Approach

> The easiest way to create a schema is through the **web interface** or by sending SQL commands using:

Reviewer: ```suggestion
The easiest way to create a schema is through the **[Web Console](/docs/web-console/)** or by sending SQL commands using:
```
